vignettes/sir_models.Rmd

```diff
@@ -0,0 +1,399 @@
+
+```{r pmcmc_combined, fig.height=5, fig.width=7}
+pmcmc_run <-
+  mcstate::pmcmc(
+    mcmc_pars,
+    filter,
+    2000,
+    save_state = TRUE,
+    save_trajectories = TRUE,
+    progress = TRUE,
+    n_chains = 1
+  )
+pmcmc_run
+plot_particle_filter(pmcmc_run$trajectories$state, true_history, incidence$day)
+```
+
+We also provide some basic tools for dealing with this output, before running prediction (see below). We can remove burn-in and thin the four chains as follows:
+
+```{r}
+processed_chains <- mcstate::pmcmc_thin(pmcmc_run, burnin = 500, thin = 2)
+parameter_mean_hpd <- apply(processed_chains$pars, 2, mean)
+parameter_mean_hpd
+```
+
+The resulting objects can also be analysed in a typical MCMC analysis package such as [`coda`](https://cran.r-project.org/web/packages/coda/index.html). We will need to combine the two parts of the object which contain the probabilities and parameter values across the iterations:
+
+```{r coda_plots, fig.height=10, fig.width=14}
+library(coda)
+mcmc1 <- coda::as.mcmc(cbind(pmcmc_run$probabilities, pmcmc_run$pars))
+
+summary(mcmc1)
+plot(mcmc1)
+```
+
+## Tuning the pMCMC
+
+The MCMC above is not particularly efficient in terms of effective sample size per iteration:
+
+```{r}
+effectiveSize(mcmc1)
+1 - rejectionRate(mcmc1)
+```
+
+Clearly the acceptance rate is too low. Generally a target of 0.5, reducing to 0.234 as the number of parameters increases, [is thought to be optimal](https://dx.doi.org/10.1016/j.spa.2007.12.005). The proposal distribution needs to be modified, in this case reducing the variance of the jump sizes. One option to automate this process is to run a chain for a bit, as above, then use the covariance of the state as the proposal matrix:
+
+```{r}
+proposal_matrix <- cov(pmcmc_run$pars)
+mcmc_pars <- mcstate::pmcmc_parameters$new(list(beta = beta, gamma = gamma), proposal_matrix)
+proposal_matrix
+```
+
+Let's now run four independent chains with these proposals:
+
+```{r, fig.height=10, fig.width=14}
+pmcmc_tuned_run <-
+  mcstate::pmcmc(
+    mcmc_pars,
+    filter,
+    2000,
+    save_state = TRUE,
+    save_trajectories = TRUE,
+    progress = TRUE,
+    n_chains = 4
+  )
+
+mcmc2 <- coda::as.mcmc(cbind(pmcmc_tuned_run$probabilities, pmcmc_tuned_run$pars))
+
+summary(mcmc2)
+plot(mcmc2)
+```
+
+Looking again at the effective sample size and rejection rate, these chains were a lot more efficient, and have a much better effective sample size thanks for the four chains:
+
+```{r}
+effectiveSize(mcmc2)
+1 - rejectionRate(mcmc2)
+```
+
+## Running predictions
+
+Going back to the simple SIR model above, we can easily continue running the fitted model forward in time using the posterior estimated by `pmcmc()` to create a forecast past the end of the data:
+
+```{r fig.height=5, fig.width=7}
+forecast <- predict(processed_chains,
+                    steps = seq(400, 800, 4),
+                    n_threads = 4L,
```

Comment:
throughout this vignette, please remove all uses of the `n_threads` argument.

The reasons for doing this are that we need to not use more than 2 threads on CRAN, but also it complicates writing the parallel vignette